### PR TITLE
Fix for #239: Union inside List or Dict is not deserialized as the correspond…

### PR DESCRIPTION
**Pull request by pawelwilczewski**
Fixed the issue described in #239.

I'm not sure this is the best way of fixing this as the comment suggests the type should have already been decoded but this is my best attempt.

Frankly, I haven't tested it properly (works for my use cases) but it seems to me the logic is valid. 

I am not very familiar with the repo, so please let me know if I'm violating some of the rules set here and I'll try to adhere.

### CODE REVIEW

**Comment by pawelwilczewski** [comment]
It would be good to add some test scenarios like the one in the mentioned issue.

**Review by george-zubrienko — COMMENTED** [review]
Hello @pawelwilczewski ! Thank you for the contribution - please add unit tests that back your claim the issue is fixed - and re-request a review ;)

**Comment by pawelwilczewski** [comment]
Stole a little bit of test code from the issue - hopefully the author a-mckinley doesn't mind

**Comment by pawelwilczewski** [comment]
@george-zubrienko done! Let me know if anything else needs changing!

**Comment by pawelwilczewski** [comment]
I noticed it won't work with `.schema.loads()` because I never made changes to the `mm.py`. I'm afraid I won't have free time to get around to doing that in the nearest future...

**Comment by pawelwilczewski** [comment]
I'd think a fix similar to what I did could be implemented in the `else` clause in line 87 in mm.py

**Comment by george-zubrienko** [comment]
> I'd think a fix similar to what I did could be implemented in the `else` clause in line 87 in mm.py

I'll review what is here this week, then we'll open a new issue and somebody can pick it up ;)

**Comment by TristanSpeakEasy** [comment]
> > I'd think a fix similar to what I did could be implemented in the `else` clause in line 87 in mm.py
> 
> I'll review what is here this week, then we'll open a new issue and somebody can pick it up ;)

would love to see this one go through as we have another PR that adds more support for undiscriminated unions we would like to go in after this https://github.com/pawelwilczewski/dataclasses-json/pull/1

**Comment by pawelwilczewski** [comment]
Not too sure why the poetry installation might have failed...

**Comment by george-zubrienko** [comment]
Merging https://github.com/lidatong/dataclasses-json/pull/476 will fix, sorry for delays in reviewing, on my todo

**Comment by pawelwilczewski** [comment]
No worries at all, and no pressure!

File: dataclasses_json/core.py [hunk 1]
```diff
@@ -310,8 +310,27 @@ def _decode_generic(type_, value, infer_missing):
...
+                for type_option in type_options:
+                    if is_dataclass(type_option):
+                        res = _decode_dataclass(type_option, value, infer_missing)
+                        changed = True
+                        break
+                if not changed:
+                    warnings.warn(
+                        f"Failed encoding {value} Union dataclasses."
+                        f"Expected Union to include a dataclass and it didn't."
+                    )
```
Comment (george-zubrienko) [inline dataclasses_json/core.py:333]:
```suggestion
            type_options = _get_type_args(type_)
            res = value
            if type(value) not dict or dict not in type_options:
                # FIXME if all types in the union are dataclasses this
                #  will just pick the first option -
                #  maybe find the best fitting class in that case instead?
                for type_option in type_options:
                    if is_dataclass(type_option):
                        res = _decode_dataclass(type_option, value, infer_missing)
                        break
                if res == value:
                    warnings.warn(
                        f"Failed to encode value {value} of type {type_}."
                        f"Expected Union to include a dataclass and it didn't."
                    )
```

**Review by george-zubrienko — DISMISSED** [review]
LGTM, minor refactoring suggestion

**Comment by pawelwilczewski** [comment]
Seems like there were some issues with the refactor. I'll roll back and you can automatically merge - unless you'd like to give it another go. Cheers!

**Comment by george-zubrienko** [comment]
I'll take a look why and suggest a new one. The one I added was from the phone so might have missed the `if` condition

**Comment by pawelwilczewski** [comment]
Yeah I gave it an attempt but kinda just skimmed through it relying on github actions rather than properly testing

**Comment by george-zubrienko** [comment]
If you can fix would be great, if not I'll prepare a proper suggestion within next 2 days. I want this to be included in 0.6.0 :)

**Comment by pawelwilczewski** [comment]
I believe this does it now

**Review by george-zubrienko — APPROVED** [review]

**Comment by george-zubrienko** [comment]
Awesome, thanks a lot for the contribution, will be released this weekend!

**Comment by pawelwilczewski** [comment]
Happy to contribute. Thanks a lot for your overall and this PR's input!

**Pull request merged**